lsp-protocol.el

```diff
@@ -0,0 +1,439 @@
+;;; lsp-protocol.el --- Language Sever Protocol Bindings  -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2020  Ivan Yonchovski
+
+;; Author: Ivan Yonchovski <yyoncho@gmail.com>
+;; Keywords: abbrev, abbrev, abbrev, abbrev
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Autogenerated bindings from lsp4j using
+;; https://github.com/victools/jsonschema-generator+scrips to generate
+;; scripts/generated.protocol.schema.json and then
+;; scripts/lsp-generate-bindings.el
+
+;;; Code:
+
+
+(defmacro lsp-interface (&rest interfaces)
+  "Generates LSP bindings from (Inteface name, Required Fields, Optional Fields) triplet.
+
+Example usage with `dash'.
+
+(-let [(&ApplyWorkspaceEditResponse
+  :failure-reason?) (ht (\"failureReason\" \"...\"))]
+  failure-reason?) "
+  (->> interfaces
+       (-map (-lambda ((interface required optional))
+               (let ((params (nconc
+                              (-map (lambda (param-name)
+                                      (cons
+                                       (intern (concat ":" (s-dashed-words (symbol-name param-name)) "?"))
+                                       param-name))
+                                    optional)
+                              (-map (lambda (param-name)
+                                      (cons (intern (concat ":" (s-dashed-words (symbol-name param-name))))
+                                            param-name))
+                                    required))))
+                 (cl-list*
+                  `(defun ,(intern (format "dash-expand:&%s" interface)) (key source)
```

Comment:
I think we're also exploring the possible to switch to use `plist` instead of hash table for better perf.
Would it make sense to generate `cl-defgeneric` and `cl-defmethod` here?
Something like
```suggestion
                  `(cl-defgeneric ,(intern (format "dash-expand:&%s" interface)) (key source))
                  `(cl-defmethod ,(intern (format "dash-expand:&%s" interface)) (key (source hash-table))
```

Reply:
> Would it make sense to generate `cl-defgeneric` and `cl-defmethod` here?
> Something like

No, because that code is executed compile time while the type info is available runtime. And it won't work either way because we should first replace all gethash/puthash with the generated getters and setters and then we will switch to plists by changing the generated code.

Reply:
> No, because that code is executed compile time while the type info is available runtime.

Can you elaborate more?
I think it will dispatch to correct method at the runtime depend on the type of source?



> And it won't work either way because we should first replace all gethash/puthash with the generated getters and setters and then we will switch to plists by changing the generated code.

I see. just think that we can just generate hash table and plist version now by `cl-defgeneric`. And require only another PR to convert everything to use generated getters/setters + set correct `json` parsing type.

Reply:
> Can you elaborate more?
I think it will dispatch to correct method at the runtime depend on the type of source?

This function is called when `dash` is expanding the `-let` and generating`let` statements during the macro expansion phase. The type of the `source` will always be `symbol`(or elisp statement). E. g.

```
(let ((foo (ht)))
  (-let [(&ClientInfo :name) foo]
    name))
```

will emit the following calls to expand function.

```
1 -> (dash-expand:&ClientInfo :name foo)
1 <- dash-expand:&ClientInfo: (gethash "name" foo)
```

Reply:
I see, thanks